Moving Folders

Folders should always have the knowable/consistant id.

If someone wants to "move" a folder, we should not update the folder record. Instead, we should create a new folder record that has the correct id generated and delete the old folder / identifier when done.

I believe we will also need to update the info in the permission/permission_reference table as well.

Notes: 
When moving/copying a folder the resulting mod_user is the system user while the expected should be the current logged in user.
when a folder contains archived items the move/copy operation gets aborted by an exception like `Caused by: com.dotmarketing.portlets.contentlet.business.DotContentletStateException: Unable to check-in an archived piece of content, please un-archive first` This behavior exists in demo. At least we should report the exception in the front end. Or fix it during move.
When reaming a folder that contains archived items the same thing happens. The check-in operation fails. However, this should be fixed by this PR  https://github.com/dotCMS/core/pull/21905

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/fileassets/business/FileAssetAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/fileassets/business/FileAssetAPIImpl.java
@@ -483,38 +483,38 @@ public class FileAssetAPIImpl implements FileAssetAPI {
 
     private boolean moveFile (Contentlet fileAssetCont, Folder parent, Host host, User user, boolean respectFrontendRoles ) throws DotStateException, DotDataException, DotSecurityException {
 
-        boolean isfileAssetContLive = false;
+		final FolderAPI folderAPI = APILocator.getFolderAPI();
 
         //Getting the contentlet identifier
         Identifier id = APILocator.getIdentifierAPI().find( fileAssetCont );
-        if ( id != null && InodeUtils.isSet( id.getId() ) ) {
+		if ( id != null && InodeUtils.isSet( id.getId() ) ) {
 
-            FileAsset fa = fromContentlet( fileAssetCont );
-            if ( fa.isLive() )
-                isfileAssetContLive = true;
+			final Folder oldParent = APILocator.getFolderAPI().findFolderByPath( id.getParentPath(), host, user, respectFrontendRoles );
+
+			final FileAsset fileAsset = fromContentlet( fileAssetCont );
 
             if ( host == null ) {
                 host = APILocator.getHostAPI().find( id.getHostId(), user, respectFrontendRoles );
             }
 
-            //Verify if the file already exist
+            //Verify if the file already exists
             Boolean fileNameExists;
             if ( parent != null ) {
-                fileNameExists = fileNameExists( host, parent, fa.getFileName(), id.getId() );
+                fileNameExists = fileNameExists( host, parent, fileAsset.getFileName(), id.getId() );
             } else {
-                fileNameExists = fileNameExists( host, APILocator.getFolderAPI().findSystemFolder(), fa.getFileName(), id.getId() );
+                fileNameExists = fileNameExists( host, APILocator.getFolderAPI().findSystemFolder(), fileAsset.getFileName(), id.getId() );
             }
 
             if ( !fileNameExists ) {
+				if(null != parent && !parent.equals(FolderAPI.SYSTEM_FOLDER)){
+					Folder parentFolder = folderAPI.find(parent.getIdentifier(), user, respectFrontendRoles);
+					id.setParentPath(parentFolder.getPath());
+				} else {
+					id.setParentPath(StringPool.FORWARD_SLASH);
+				}
+				id.setHostId(host != null ? host.getIdentifier() : (parent != null ? parent.getHostId() : fileAssetCont.getHost()) );
+				id = identifierAPI.save(id);
 
-                Folder oldParent = APILocator.getFolderAPI().findFolderByPath( id.getParentPath(), host, user, respectFrontendRoles );
-
-                fileAssetCont.setInode( null );
-                fileAssetCont.setHost( host != null ? host.getIdentifier() : (parent != null ? parent.getHostId() : fileAssetCont.getHost()) );
-                fileAssetCont.setFolder( parent != null ? parent.getInode() : null );
-                final Contentlet fileAsset = APILocator.getContentletAPI().checkin( fileAssetCont, user, respectFrontendRoles );
-                if ( isfileAssetContLive )
-                    APILocator.getVersionableAPI().setLive( fileAssetCont );
 
                 if ( parent != null ) {
                     CacheLocator.getNavToolCache().removeNav(parent.getHostId(), parent.getInode());
@@ -522,9 +522,7 @@ public class FileAssetAPIImpl implements FileAssetAPI {
 
                 CacheLocator.getNavToolCache().removeNav(oldParent.getHostId(), oldParent.getInode());
 
-                CacheLocator.getIdentifierCache().removeFromCacheByVersionable( fileAssetCont );
-
-				this.systemEventsAPI.pushAsync(SystemEventType.MOVE_FILE_ASSET, new Payload(fileAsset.getMap(), Visibility.EXCLUDE_OWNER,
+				this.systemEventsAPI.pushAsync(SystemEventType.MOVE_FILE_ASSET, new Payload(fileAssetCont.getMap(), Visibility.EXCLUDE_OWNER,
 					new ExcludeOwnerVerifierBean(user.getUserId(), PermissionAPI.PERMISSION_READ, Visibility.PERMISSION)));
 
                 return true;

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/fileassets/business/FileAssetAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/fileassets/business/FileAssetAPIImpl.java
@@ -2,14 +2,12 @@ package com.dotmarketing.portlets.fileassets.business;
 
 import com.dotcms.api.tree.Parentable;
 import com.dotcms.browser.BrowserQuery;
-import com.dotmarketing.db.HibernateUtil;
 import com.dotmarketing.portlets.structure.model.Field.DataType;
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
-import java.net.ConnectException;
 import java.nio.file.Files;
 import java.util.ArrayList;
 import java.util.List;
@@ -17,7 +15,6 @@ import java.util.Map;
 import java.util.stream.Collectors;
 import java.util.zip.GZIPInputStream;
 
-import com.dotcms.exception.ExceptionUtil;
 import com.dotcms.repackage.com.google.common.annotations.VisibleForTesting;
 import com.dotcms.util.MimeTypeUtils;
 import com.dotmarketing.business.*;
@@ -30,7 +27,6 @@ import com.dotcms.api.system.event.Visibility;
 import com.dotcms.api.system.event.verifier.ExcludeOwnerVerifierBean;
 import com.dotcms.business.CloseDBIfOpened;
 import com.dotcms.business.WrapInTransaction;
-import com.dotcms.contenttype.model.type.BaseContentType;
 import com.dotcms.rendering.velocity.viewtools.content.FileAssetMap;
 import com.dotcms.repackage.org.apache.commons.io.IOUtils;
 import com.dotcms.tika.TikaUtils;
@@ -40,7 +36,6 @@ import com.dotmarketing.cache.FieldsCache;
 import com.dotmarketing.exception.DotDataException;
 import com.dotmarketing.exception.DotRuntimeException;
 import com.dotmarketing.exception.DotSecurityException;
-import com.dotmarketing.menubuilders.RefreshMenus;
 import com.dotmarketing.portlets.contentlet.business.ContentletAPI;
 import com.dotmarketing.portlets.contentlet.business.DotContentletStateException;
 import com.dotmarketing.portlets.contentlet.model.Contentlet;
@@ -58,7 +53,6 @@ import com.liferay.portal.model.User;
 import com.liferay.util.FileUtil;
 import com.liferay.util.StringPool;
 import io.vavr.control.Try;
-import org.apache.velocity.exception.ResourceNotFoundException;
 
 /**
  * This class is a bridge impl that will support the older
@@ -483,40 +477,31 @@ public class FileAssetAPIImpl implements FileAssetAPI {
 
     private boolean moveFile (Contentlet fileAssetCont, Folder parent, Host host, User user, boolean respectFrontendRoles ) throws DotStateException, DotDataException, DotSecurityException {
 
-		final FolderAPI folderAPI = APILocator.getFolderAPI();
-
         //Getting the contentlet identifier
         Identifier id = APILocator.getIdentifierAPI().find( fileAssetCont );
-		if ( id != null && InodeUtils.isSet( id.getId() ) ) {
+        if ( id != null && InodeUtils.isSet( id.getId() ) ) {
 
-			final Folder oldParent = APILocator.getFolderAPI().findFolderByPath( id.getParentPath(), host, user, respectFrontendRoles );
-
-			final FileAsset fileAsset = fromContentlet( fileAssetCont );
+            final FileAsset fa = fromContentlet( fileAssetCont );
 
             if ( host == null ) {
                 host = APILocator.getHostAPI().find( id.getHostId(), user, respectFrontendRoles );
             }
 
-            //Verify if the file already exists
+            //Verify if the file already exist
             Boolean fileNameExists;
             if ( parent != null ) {
-                fileNameExists = fileNameExists( host, parent, fileAsset.getFileName(), id.getId() );
+                fileNameExists = fileNameExists( host, parent, fa.getFileName(), id.getId() );
             } else {
-                fileNameExists = fileNameExists( host, APILocator.getFolderAPI().findSystemFolder(), fileAsset.getFileName(), id.getId() );
+                fileNameExists = fileNameExists( host, APILocator.getFolderAPI().findSystemFolder(), fa.getFileName(), id.getId() );
             }
 
             if ( !fileNameExists ) {
-				if(null != parent && !parent.equals(FolderAPI.SYSTEM_FOLDER)){
-					Folder parentFolder = folderAPI.find(parent.getIdentifier(), user, respectFrontendRoles);
-					id.setParentPath(parentFolder.getPath());
-				} else {
-					id.setParentPath(StringPool.FORWARD_SLASH);
-				}
-				id.setHostId(host != null ? host.getIdentifier() : (parent != null ? parent.getHostId() : fileAssetCont.getHost()) );
-				id = identifierAPI.save(id);
 
+				APILocator.getContentletAPI().move(fileAssetCont, user, host, parent, respectFrontendRoles);
 
-                if ( parent != null ) {
+				final Folder oldParent = APILocator.getFolderAPI().findFolderByPath( id.getParentPath(), host, user, respectFrontendRoles );
+
+				if ( parent != null ) {
                     CacheLocator.getNavToolCache().removeNav(parent.getHostId(), parent.getInode());
                 }
 

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderFactoryImpl.java
@@ -673,25 +673,16 @@ public class FolderFactoryImpl extends FolderFactory {
 									  final User systemUser,
 									  final List<Contentlet> contentlets) throws DotDataException, DotSecurityException {
 
-		final ContentletAPI contentletAPI = APILocator.getContentletAPI();
-		final FileAssetAPI  fileAssetAPI  = APILocator.getFileAssetAPI();
-
 		for(final Contentlet contentlet : contentlets) {
-
 			if (contentlet.isFileAsset()) {
+				final FileAssetAPI  fileAssetAPI  = APILocator.getFileAssetAPI();
 				fileAssetAPI.moveFile(contentlet, folder, systemUser, false);
 		    } else if (contentlet.isHTMLPage()) {
 				HTMLPageAssetAPI pageAssetAPI = APILocator.getHTMLPageAssetAPI();
 				pageAssetAPI.move(pageAssetAPI.fromContentlet(contentlet), folder, systemUser);
 			} else {
-    			final boolean isLive = contentlet.isLive();
-				Contentlet newContentlet  = contentletAPI.checkout(contentlet.getInode(), systemUser, false);
-				newContentlet.setFolder(folder.getInode());
-				newContentlet = contentletAPI.checkin(newContentlet, systemUser, false);
-    			if(isLive) {
-
-					contentletAPI.publish(newContentlet, systemUser, false);
-    			}
+				final ContentletAPI contentletAPI = APILocator.getContentletAPI();
+				contentletAPI.move(contentlet, systemUser, folder.getHost(), folder, false);
 		    }
 		}
 	}

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/HTMLPageAssetAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/HTMLPageAssetAPIImpl.java
@@ -34,7 +34,6 @@ import com.dotmarketing.filters.CMSUrlUtil;
 import com.dotmarketing.portlets.contentlet.business.ContentletAPI;
 import com.dotmarketing.portlets.contentlet.model.Contentlet;
 import com.dotmarketing.portlets.contentlet.model.ContentletVersionInfo;
-import com.dotmarketing.portlets.folders.business.FolderAPI;
 import com.dotmarketing.portlets.folders.model.Folder;
 import com.dotmarketing.portlets.htmlpageasset.model.HTMLPageAsset;
 import com.dotmarketing.portlets.htmlpageasset.model.IHTMLPage;
@@ -54,10 +53,8 @@ import com.dotmarketing.util.UtilMethods;
 import com.dotmarketing.util.WebKeys;
 import com.google.common.collect.ImmutableList;
 import com.liferay.portal.model.User;
-import com.liferay.util.StringPool;
 import java.util.ArrayList;
 import java.util.Date;
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
@@ -423,7 +420,7 @@ public class HTMLPageAssetAPIImpl implements HTMLPageAssetAPI {
     }
 
     /**
-     * @see HTMLPageAssetAPI#findPage(inode, User, boolean)
+     * @see HTMLPageAssetAPI#findPage(String, User, boolean)
      */
     @CloseDBIfOpened
     @Override
@@ -499,14 +496,18 @@ public class HTMLPageAssetAPIImpl implements HTMLPageAssetAPI {
         Identifier targetIdent = identifierAPI.find(host,
                 targetFolderIdent.getURI() + sourceIdent.getAssetName());
         if (targetIdent == null || !InodeUtils.isSet(targetIdent.getId())) {
+            final Contentlet contentlet = contentletAPI
+                    .find(page.getInode(), user, false);
 
-            if(null != parent && !parent.equals(FolderAPI.SYSTEM_FOLDER)){
-                sourceIdent.setParentPath(targetFolderIdent.getPath());
-            } else {
-                sourceIdent.setParentPath(StringPool.FORWARD_SLASH);
+            contentletAPI.move(contentlet, user, host, parent, false);
+
+            if ( parent != null ) {
+                CacheLocator.getNavToolCache().removeNav(parent.getHostId(), parent.getInode());
             }
-            sourceIdent.setHostId(host != null ? host.getIdentifier() : (parent != null ? parent.getHostId() : page.getHost()) );
-            identifierAPI.save(sourceIdent);
+
+            final Folder oldParent = APILocator.getFolderAPI().findFolderByPath( sourceIdent.getParentPath(), host, user, false);
+
+            CacheLocator.getNavToolCache().removeNav(oldParent.getHostId(), oldParent.getInode());
 
             systemEventsAPI.pushAsync(SystemEventType.MOVE_PAGE_ASSET, new Payload(page.getMap(), Visibility.EXCLUDE_OWNER,
                     new ExcludeOwnerVerifierBean(user.getUserId(), PermissionAPI.PERMISSION_READ, Visibility.PERMISSION)));

--- a/dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/HTMLPageAssetAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/HTMLPageAssetAPIImpl.java
@@ -34,6 +34,7 @@ import com.dotmarketing.filters.CMSUrlUtil;
 import com.dotmarketing.portlets.contentlet.business.ContentletAPI;
 import com.dotmarketing.portlets.contentlet.model.Contentlet;
 import com.dotmarketing.portlets.contentlet.model.ContentletVersionInfo;
+import com.dotmarketing.portlets.folders.business.FolderAPI;
 import com.dotmarketing.portlets.folders.model.Folder;
 import com.dotmarketing.portlets.htmlpageasset.model.HTMLPageAsset;
 import com.dotmarketing.portlets.htmlpageasset.model.IHTMLPage;
@@ -53,6 +54,7 @@ import com.dotmarketing.util.UtilMethods;
 import com.dotmarketing.util.WebKeys;
 import com.google.common.collect.ImmutableList;
 import com.liferay.portal.model.User;
+import com.liferay.util.StringPool;
 import java.util.ArrayList;
 import java.util.Date;
 import java.util.HashMap;
@@ -497,35 +499,14 @@ public class HTMLPageAssetAPIImpl implements HTMLPageAssetAPI {
         Identifier targetIdent = identifierAPI.find(host,
                 targetFolderIdent.getURI() + sourceIdent.getAssetName());
         if (targetIdent == null || !InodeUtils.isSet(targetIdent.getId())) {
-            Contentlet contentlet = contentletAPI
-                    .find(page.getInode(), user, false);
 
-            /*
-             * Enable multi-language support when cut a page
-             */
-            Map<String, Boolean> inodesToMove = new HashMap<String, Boolean>();
-            // Getting all working contentlet version languages
-            for(Contentlet workingContentlet : contentletAPI.getAllLanguages(contentlet, false,
-                    user, false)) {
-                inodesToMove.put(workingContentlet.getInode(), false);
+            if(null != parent && !parent.equals(FolderAPI.SYSTEM_FOLDER)){
+                sourceIdent.setParentPath(targetFolderIdent.getPath());
+            } else {
+                sourceIdent.setParentPath(StringPool.FORWARD_SLASH);
             }
-            // Getting all live contentlet version languages
-            for(Contentlet liveContentlet : contentletAPI.getAllLanguages(contentlet, true,
-                    user, false)) {
-                inodesToMove.put(liveContentlet.getInode(), true);
-            }
-
-            for (String contentletInode : inodesToMove.keySet()) {
-                Contentlet cont = contentletAPI.checkout(contentletInode, user, false);
-
-                cont.setFolder(parent.getInode());
-                cont.setHost(host.getIdentifier());
-                cont = contentletAPI.checkin(cont, user, false);
-                if (inodesToMove.get(contentletInode)) {
-                    // We need to publish those that were live
-                    contentletAPI.publish(cont, user, false);
-                }
-            }
+            sourceIdent.setHostId(host != null ? host.getIdentifier() : (parent != null ? parent.getHostId() : page.getHost()) );
+            identifierAPI.save(sourceIdent);
 
             systemEventsAPI.pushAsync(SystemEventType.MOVE_PAGE_ASSET, new Payload(page.getMap(), Visibility.EXCLUDE_OWNER,
                     new ExcludeOwnerVerifierBean(user.getUserId(), PermissionAPI.PERMISSION_READ, Visibility.PERMISSION)));

--- a/dotCMS/src/main/webapp/html/portlet/ext/folders/edit_folder.jsp
+++ b/dotCMS/src/main/webapp/html/portlet/ext/folders/edit_folder.jsp
@@ -39,7 +39,7 @@ dojo.require("dotcms.dojo.data.StructureReadStore");
 
 	var formName;
 	function save(formName) {
-
+        dijit.byId('processingDialog').show();
 		this.formName = formName;
 
 		var form = document.getElementById(this.formName);
@@ -266,6 +266,10 @@ dojo.require("dotcms.dojo.data.StructureReadStore");
 	%>
 	<!-- END permissions -->
 	</div>
+</div>
+
+<div id="processingDialog" dojoType="dijit.Dialog" disableCloseButton="true" title="<%=LanguageUtil.get(pageContext,"Processing")%>" style="display: none;">
+    <div dojoType="dijit.ProgressBar" style="width:200px;text-align:center;" indeterminate="true" jsId="processingLoading" id="processingLoading"></div>
 </div>
 
 </liferay:box>

